### ISSUE TITLE
Kademlia: NAT / don't announce private address and high peers churn

### ISSUE THREAD

**Issue by BlackYoup** [libp2p/rust-libp2p#1771]
Hello, hope everyone is well!

I have a few questions / remarks on how to use kademlia with a NATed network (even though all peers are publicly reachable) and where the churn of peers might be high. I hesitated to open two separate issues but since they share the same context I did not. If you think those are valid problems and not related, then I'll open separate issues :) 

I'll describe my use case first and my current issues.

I'm working at Clever Cloud and we basically do application hosting. The goal of our project is to have a peer2peer daemon on each application's instances to share dependencies between peers. Every instances are in the same DHT and may contain ten of thousands of peers.
Each time an application is deployed, a certain amount of new instances are created and another amount of instances are stopped. Each instance will publish providers records for a certain amount of dependencies (could be 10 or 10 000). Instances that are in need for dependencies will ask the providers of a record in the DHT. Instances are also behind a NAT but they are reachable using a public port. For example, on the instance, the daemon will listen on address `10.0.0.1:6060` but on the outside, it will be reachable using `169.0.0.1:10000` (`169.0.0.1` is to be replaced by a public IP).

Note that each peer knows its public address (IP + port), the address is registered using `Swarm::add_external_address()` at the start of the daemon.

Now, I'm having two separate issues / questions:

## Kademlia keeps trying to connect to peers it discovers even though previous attempts failed

This may be expected behavior. My problem here is that when an instance is stopped, it will never be up again which means that if we loose the connection to a remote peer or can't connect to it, it means it has been stopped (99% of the time, 1% remaining might be network issues for example but I'm ignoring those for now).

Here, I saw that becoming an issue in the following scenario:

- 1 bootstrap node (A) always available, every peer connect to it to join the DHT
- 2 peers (B and C) joined the DHT and start providing records
- Those 2 peers (B and C) are stopped
- 2 new peers (D and E) start and join the DHT

`D` and `E` will then ask the DHT if it knows about some records. The bootstrap node will forward peers it knows about which are `B` and `C` peers, currently down. `D` and `E` will then try to connect to them, see they are down and won't have a response, life goes on.

`D` and `E` then do a second DHT request. `A` will again forward peers it knows which are still `B` and `C`. `D` and `E` will try to connect again to those peers and fail. And this goes on, also leading to a lot of timeouts in kademlia requests (my timeout trigger is quite low, 5 seconds).

Is there a way currently in kademlia to avoid this? From reading the code it doesn't seem so but would it even be feasible?

## Kademlia announces local IP of the peer

If I read the code correctly, kademlia announces the addresses of the local peer by adding the local addresses to the query response. In my case, it makes kademlia announce the internal address (`10.0.0.1:6060`) among other addresses, which is not reachable by any other peer. Other peers may try to connect to this IP and fail with a `no route to host` error. As you may know, this error is slower to "trigger" than a classic `connection refused`. Please note that the internal address in not the first one to be tried to connect to this node. Usually the external address is tried before but fails and then kademlia goes on and try every other addresses is has recorded.

Ideally for my use case, only the public address of the peer would be announced and kademlia would only try to connect to this public address. I couldn't see a way to remove an address from the local address of the node, would it make sense? Is there a better way to avoid this issue?

This, combined with the issue I mentioned above, makes the daemon quite slow (x19) compared to when all peers are correctly connected mostly because it always fail to connect to the internal address and keep retrying.

Thank you for any input you could provide on those issues!

**Comment by mxinden** [comment]
Hi @BlackYoup,

This sounds like a cool use-case. Keep us posted!

### A continues returning offline nodes B and C

The kbucket eviction strategy used in this implementation is lazy. It does not actively probe whether nodes in its kbuckets are offline, but only evicts a node if it is either explicitly removed via `remove_address` / `remove_peer`, or implicitly failed to respond to some other action, e.g. a local query, injected via `inject_addr_reach_failure`.

Also see the ["Optimized contact accounting" in the original Kademlia paper](https://pdos.csail.mit.edu/~petar/papers/maymounkov-kademlia-lncs.pdf).

> Is there a way currently in kademlia to avoid this? From reading the code it doesn't seem so but would it even be feasible?

As of today there is no way to actively ensure peers in ones kbuckets are still alive. As a hack you could regularly ping each peer in the Kademlia kbuckets.

I would expect this to be an issue of small testing networks only, as within larger networks offline peers would eventually be pushed out of kbuckets by online peers. Maybe you can extend on your use case if you think differently.


### Local IP addresses within DHT

As far as I know there is no way to prevent a Kademlia instance to advertise its own local listening address. Instead you can manually control which addresses of remote peers are added to the local routing table (kbuckets) via the [`KademliaBucketInserts` strategy](https://github.com/libp2p/rust-libp2p/blob/master/protocols/kad/src/behaviour.rs#L102) thus allowing you to filter out non-global remote addresses. We do this within [Substrate](https://github.com/paritytech/substrate/blob/master/client/network/src/discovery.rs#L96-L97).

Again as an intermediary hack you could implement a transport wrapper discarding any non-global addresses, see [here](https://github.com/mxinden/kademlia-exporter/blob/master/src/exporter/client/global_only.rs).

Let me know if the above is of some help.

**Comment by BlackYoup** [comment]
Hello!

Thanks for your reply :)

> As of today there is no way to actively ensure peers in ones kbuckets are still alive. As a hack you could regularly ping each peer in the Kademlia kbuckets.

That's a great idea, I didn't think about it, I'll give it a try :+1: 

> I would expect this to be an issue of small testing networks only, as within larger networks offline peers would eventually be pushed out of kbuckets by online peers. Maybe you can extend on your use case if you think differently.

:+1: 

>  Instead you can manually control which addresses of remote peers are added to the local routing table

Yeah, I saw the `KademliaBucketInserts` strategy but failed to see how I could use it for this particular issue. I'll look into your examples then.

Thanks a lot!

**Comment by BlackYoup** [comment]
Hi @mxinden ,

I've tested a few thing related to your comments, here's what I have now:

> Instead you can manually control which addresses of remote peers are added to the local routing table

I implemented my own "identify" messages so that a peer can announce its public IP to another peer once they are connected. Once a peer knows the public IP of the other peer, it adds it in the kademlia routing table. Thanks again for this pointer :+1: 

> As of today there is no way to actively ensure peers in ones kbuckets are still alive. As a hack you could regularly ping each peer in the Kademlia kbuckets.

I implemented a small set of ping messages where each node pings each other. If we fail to send the message or do not have a response in time, then I consider this peer down and remove it from the kademlia routing table `kademlia::remove_peer`).

The issue that remains is that when a peer disconnects, kademlia still seems to announce it to other peers, even though I called the `kademlia::remove_peer` method which returns `None`. I understand why a node that disconnects could still be announced but I would have expected the `remove_peer` method to make kademlia "forget" about this peer once and for all.

Is there anything I'm missing here?

> I would expect this to be an issue of small testing networks only

Thanks for you kademlia paper link you gave, I now understand better the eviction of peers. Unfortunately, we probably will deploy our solution on large networks but also on small networks. And even at the start of the deployment process, the DHT will be of a few nodes to "test" in production environment and see if everything goes smoothly, before deploying it wider. So I'll need it to work even for small networks.

By the way, the `RequestResponse` network behaviour is a nice addition, thank you :)

**Comment by romanb** [comment]
> The issue that remains is that when a peer disconnects, kademlia still seems to announce it to other peers, even though I called the kademlia::remove_peer method which returns None. I understand why a node that disconnects could still be announced but I would have expected the remove_peer method to make kademlia "forget" about this peer once and for all.

That should be the case, unless there is bug in `remove_peer`. Of course, this is a local operation and removes the peer from the _local_ routing table. When a peer disappears from the network and doesn't come back, it may still take some time for all nodes to notice and remove that peer from that routing table. Until that is the case, iterative queries can still discover this peer from some node's routing table and especially in small networks every query will probably still "discover" that peer as long as just one node still has it in its routing table. It won't be added back to the routing table unless explicitly done via `add_address`, since connection attempts fail. If you could describe a specific situation or series of events that does not seem right, please let us know.

> I couldn't see a way to remove an address from the local address of the node, would it make sense? Is there a better way to avoid this issue?

I assume you're referring to [this code](https://github.com/libp2p/rust-libp2p/blob/master/protocols/kad/src/behaviour.rs#L1044-L1053) relating to `ADD_PROVIDER` requests? Since I think all other addresses that circulate come from the routing table and can be fully controlled via `add_address` and `remove_address`, as @mxinden mentioned in the context of `KademliaBucketInserts::Manual`. I guess right now the behaviour in which the `Kademlia` behaviour is contained could pass down its own `impl PollParameters` to `<Kademlia as NetworkBehaviour>::poll()` and thus filter the "external addresses" visible to `Kademlia` (unless you're using the `derive` macro maybe). However, it may indeed be sensible to make a simple configuration option for not announcing addresses in private address ranges, as these are mostly useful for test or development setups.

**Comment by BlackYoup** [comment]
> That should be the case, unless there is bug in remove_peer

Yes, I understand that if a node didn't remove the peer from its local routing table, it might continue to announce it. But I can trigger a node announcing a peer after this node called `remove_peer` for this particular peer. I'll open a separate issue for this matter as I can reproduce it quite easily :+1:

 > I assume you're referring to this code relating to ADD_PROVIDER requests?

Yes! Indeed, manually adding the peers solved this issue. I'll give a try to the option to avoid announcing private addresses.

**Comment by BlackYoup** [comment]
As a small heads up, I've managed to fix both of the issues described here.

About kademlia that continuously discovers peers that were received in the `GetProviders` response, I made a wrapper around the `MemoryStore` struct and I overloaded the `providers()` method to only return the providers that are not known to be down. This way, nodes answering a kademlia query won't forward the providers for a record that they know to be down. This and the continuous ping of the peers in the kbucket seems to fix the issue.

I'll see how this goes in the future.

Thanks again for your help here.
